build.rs

```diff
@@ -25,9 +41,473 @@ struct CgDevice {
+                Some(ParserState::HidType(Map::<u8>::new(), None))
+            }
+            "# R ite" => {
+                self.finalize(output);
+                Some(ParserState::RType(Map::<u8>::new(), None))
+            }
+            "# BIAS " => {
+                self.finalize(output);
+                Some(ParserState::BiasType(Map::<u8>::new(), None))
+            }
+            "# PHY i" => {
+                self.finalize(output);
+                Some(ParserState::PhyType(Map::<u8>::new(), None))
+            }
+            "# HUT h" => {
+                self.finalize(output);
+                Some(ParserState::HutType(Map::<u8>::new(), None))
+            }
+            "# L lan" => {
+                self.finalize(output);
+                Some(ParserState::Lang(Map::<u16>::new(), None))
+            }
+            "# HCC c" => {
+                self.finalize(output);
+                Some(ParserState::CountryCode(Map::<u8>::new(), None))
+            }
+            "# VT te" => {
+                self.finalize(output);
+                Some(ParserState::TerminalType(Map::<u16>::new(), None))
+            }
+            _ => None,
+        }
+    }
+
+    /// Process a line of input for the current state
+    fn process(&mut self, line: &str) {
+        if line.is_empty() || line.starts_with('#') {
+            return;
+        }
+
+        // Switch parser state based on line prefix and current state
+        // this relies on ordering of classes and types in the file...
+        match self {
+            ParserState::Vendors(m, ref mut curr_vendor, ref mut curr_device_id) => {
+                if let Ok((name, id)) = parser::vendor(line) {
+                    if let Some(cv) = curr_vendor {
+                        m.entry(cv.id, &quote!(#cv).to_string());
+                    }
+
+                    // Set our new vendor as the current vendor.
+                    *curr_vendor = Some(CgVendor {
+                        id,
+                        name: name.into(),
+                        devices: vec![],
+                    });
+                // We should always have a current vendor; failure here indicates a malformed input.
+                } else if let Some(curr_vendor) = curr_vendor.as_mut() {
+                    if let Ok((name, id)) = parser::device(line) {
+                        curr_vendor.devices.push(CgDevice {
+                            id,
+                            name: name.into(),
+                            interfaces: vec![],
+                        });
+                        *curr_device_id = id;
+                    } else if let Ok((name, id)) = parser::interface(line) {
+                        let curr_device = curr_vendor
+                            .devices
+                            .iter_mut()
+                            .find(|d| d.id == *curr_device_id)
+                            .expect("No parent device whilst parsing interfaces, confirm file not malformed");
```

Comment:
Nitpick, but maybe make this and similar errors a bit shorter? The error state implies a malformed input and should really only happen on the CI here (since we only ship copies that we know work), so we can afford to be a little terse :slightly_smiling_face: 

```suggestion
                            .expect("No parent device whilst parsing interfaces");
```

Reply:
(And similar for the other error strings)